libsolidity/ExpressionCompiler.cpp

```diff
@@ -628,7 +628,41 @@ void ExpressionCompiler::appendTypeConversion(Type const& _typeOnStack, Type con
 		return;
 	Type::Category stackTypeCategory = _typeOnStack.getCategory();
 	Type::Category targetTypeCategory = _targetType.getCategory();
-	if (stackTypeCategory == Type::Category::INTEGER || stackTypeCategory == Type::Category::CONTRACT ||
+
+	if (stackTypeCategory == Type::Category::STRING)
+	{
+		if (targetTypeCategory == Type::Category::INTEGER)
+		{
+			// conversion from string to hash. no need to clean the high bit
+			// only to shift right because of opposite alignment
+			IntegerType const& targetIntegerType = dynamic_cast<IntegerType const&>(_targetType);
+			StaticStringType const& sourceStringType = dynamic_cast<StaticStringType const&>(_typeOnStack);
+			if (targetIntegerType.isHash())
+			{
+				solAssert(targetIntegerType.getNumBits() == sourceStringType.getNumBytes() * 8, "The size should be the same.");
+				m_context << (u256(1) << 256 - sourceStringType.getNumBytes() * 8) <<
+						eth::Instruction::SWAP1 << eth::Instruction::DIV;
+			}
+		}
+		else {
+			solAssert(targetTypeCategory == Type::Category::STRING, "Invalid type conversion requested.");
+			// nothing to do, strings are high-order-bit-aligned
+			//@todo clear lower-order bytes if we allow explicit conversion to shorter strings
+		}
+	}
+	else if (targetTypeCategory == Type::Category::STRING && stackTypeCategory == Type::Category::INTEGER)
```

Comment:
Please add this case as a subcase of `stackTypeCategory == Type::Category::INTEGER`.
Comments from above also apply here.